### DISCUSSION TITLE
Will MinIO consider to support additional checksum algorithms for amazon s3?

### DISCUSSION THREAD

**Discussion by moting9** [minio/minio#15723] [Q&A] [answered]
Hello:

 MD5 takes relatively high CPU usage at object PUT.
Amazon mentioned it supports additional checksum algorithms besides MD5, such as crc-32, crc-32c, sha-1, sha-256

https://aws.amazon.com/cn/blogs/aws/new-additional-checksum-algorithms-for-amazon-s3/

Just a query, will MinIO consider to support such options? Thanks!

**Comment by klauspost** [comment, marked as answer]
This is supported from a few releases back in #15433 

Trailing checksums are currently being implemented.

**Reply by adam-cashman** [reply]
Hello,
I see that there is a draft PR for trailing checksums ( #16484 ). Do you have an expectation as to when that may be merged / released? (I apologize in advance for not being more familiar with the project's typical release process/cycle). Thanks!

**Comment by moting9** [comment]
Thanks klauspost!  
MD5 is calculated  for Etag in response header in server side.
https://docs.aws.amazon.com/AmazonS3/latest/API/RESTCommonResponseHeaders.html
https://docs.aws.amazon.com/AmazonS3/latest/API/API_Object.html

I understand that the additonal checksum feature may benefit s3 client to reduce hash calculation workload, MD5 calcuation for Etag in server side is still CPU intensive.

**Reply by harshavardhana** [reply]
md5 cannot be turned-off @moting9 - that will break S3 compatibility.

**Reply by moting9** [reply]
Understand. Thanks!

**Reply by harshavardhana** [reply]
if you want to experiment you can use `minio server --no-compat` mode where md5sum is turned-off

**Reply by moting9** [reply]
Hello, harshavardhana, 
I studied latest MinIO code, no-campat is for skipsha256cksum, not for md5. 
no-campat ->  globalCLIContext.StrictS3Compat -> skipContentSha256Cksum

I raised the question is becuase that  MD5 is heavily used in object storage PUT,  Now I understand that the  "additonal checksum" feature by Amazon will not change Etag calcuation method in server side. 
 
I don't quite understand how Etag is used in object storage for client/server to sync data integrity. 
Tencent COS uses "x-cos-hash-crc64ecma" instead of MD5 for Etag calculation, seems it ever meet with data migration issue because Etag is not MD5 format.   https://github.com/rclone/rclone/issues/4094

**Reply by harshavardhana** [reply]
ETag is not always md5 @moting9
